llarp/messages/relay_status.cpp

```diff
@@ -0,0 +1,295 @@
+      {
+        return false;
+      }
+    }
+    else if(key == "v")
+    {
+      if(!BEncodeMaybeReadVersion("v", version, LLARP_PROTO_VERSION, read, key,
+                                  buf))
+      {
+        return false;
+      }
+    }
+
+    return read;
+  }
+
+  void
+  LR_StatusMessage::Clear()
+  {
+    frames[0].Clear();
+    frames[1].Clear();
+    frames[2].Clear();
+    frames[3].Clear();
+    frames[4].Clear();
+    frames[5].Clear();
+    frames[6].Clear();
+    frames[7].Clear();
+  }
+
+  bool
+  LR_StatusMessage::BEncode(llarp_buffer_t* buf) const
+  {
+    if(!bencode_start_dict(buf))
+      return false;
+    // msg type
+    if(!BEncodeWriteDictMsgType(buf, "a", "s"))
+      return false;
+    // frames
+    if(!BEncodeWriteDictArray("c", frames, buf))
+      return false;
+    // path id
+    if(!BEncodeWriteDictEntry("p", pathid, buf))
+      return false;
+    // status (for now, only success bit is relevant)
+    if(!BEncodeWriteDictInt("s", status, buf))
+      return false;
+    // version
+    if(!bencode_write_version_entry(buf))
+      return false;
+
+    return bencode_end(buf);
+  }
+
+  bool
+  LR_StatusMessage::HandleMessage(AbstractRouter* router) const
+  {
+    llarp::LogDebug("Received LR_Status message from (", session->GetPubKey(), ")");
+    if(frames.size() != path::max_len)
+    {
+      llarp::LogError("LRSM invalid number of records, ", frames.size(),
+                      "!=", path::max_len);
+      return false;
+    }
+
+    auto path = router->pathContext().GetByUpstream(session->GetPubKey(), pathid);
+    if (!path)
+    {
+      llarp::LogWarn("unhandled LR_Status message: no associated IHopHandler found");
+      return false;
+    }
+
+    auto handler = std::make_shared< LRSM_AsyncHandler >(
+        frames, pathid, status, path, router);
+
+    handler->queue_handle();
+
+    return true;
+  }
+
+  void LR_StatusMessage::SetDummyFrames()
+  {
+    //TODO
+    return;
+  }
+
+  // call this from a worker thread
+  bool
+  LR_StatusMessage::CreateAndSend(AbstractRouter *router, const PathID_t pathid,
+                                  const RouterID nextHop, const SharedSecret pathKey,
+                                  uint64_t status)
+  {
+    typedef std::shared_ptr< LR_StatusMessage > LRSM_ptr;
+    LRSM_ptr message = std::make_shared< LR_StatusMessage >();
+
+    message->status = status & LR_StatusRecord::SUCCESS;
+    message->pathid = pathid;
+
+    message->SetDummyFrames();
+
+    if(!message->AddFrame(pathKey, status))
+    {
+      return false;
+    }
+
+    QueueSendMessage(router, nextHop, message);
+    return true; // can't guarantee delivery here, as far as we know it's fine
+  }
+
+  bool
+  LR_StatusMessage::AddFrame(const SharedSecret& pathKey, uint64_t status)
+  {
+    frames[7] = frames[6];
+    frames[6] = frames[5];
+    frames[5] = frames[4];
+    frames[4] = frames[3];
+    frames[3] = frames[2];
+    frames[2] = frames[1];
+    frames[1] = frames[0];
```

Comment:
I think this is equivalent:

```
std::copy(frames.begin(), frames.end() - 1, frames.begin() + 1);
auto& frame = frames[0];
```

Reply:
probably, will have to consider which reads better.  not too worried about a few extra lines.

I assume they would compile to equivalently efficient code, but that may be a poor assumption.

Reply:
kept as-is for now, still not convinced the brevity here is worth the change in readability, but perhaps this is a result of my not having used std::copy very much.